src/OpenLoco/Windows/Vehicle.cpp

```diff
@@ -2265,13 +2265,129 @@ namespace OpenLoco::Ui::Vehicle
+
+                _F2494A[i].orderOffset = order.getOffset();
+
+                auto lineNumber = 0;
+                auto* stationOrder = order.as<Vehicles::OrderStation>();
+                auto* waypointOrder = order.as<Vehicles::OrderRouteWaypoint>();
+                if (stationOrder != nullptr)
+                {
+                    lineNumber++; // station labels start on line 0 so add 1 for the number
+
+                    const auto stationId = stationOrder->getStation();
+                    for (auto innerOrder = orders.begin(); innerOrder->getOffset() != order.getOffset(); ++innerOrder)
+                    {
+                        auto* innerStationOrder = innerOrder->as<Vehicles::OrderStation>();
+                        if (innerStationOrder == nullptr)
+                        {
+                            continue;
+                        }
+                        if (stationId == innerStationOrder->getStation())
+                        {
+                            lineNumber++;
+                        }
+                    }
+                }
+                else if (waypointOrder != nullptr)
+                {
+                    const auto rawOrder = order.getRaw();
+                    for (auto innerOrder = orders.begin(); innerOrder->getOffset() != order.getOffset(); ++innerOrder)
+                    {
+                        if (rawOrder == innerOrder->getRaw())
+                        {
+                            lineNumber++;
+                        }
+                    }
+                }
+
+                _F2494A[i].lineNumber = lineNumber & 0xFF;
+                i++;
+            }
+
+            i = 0;
+            for (auto& unk : _F2494A)
+            {
+                if (unk.orderOffset == static_cast<uint32_t>(-1))
+                {
+                    continue;
+                }
+
+                auto order = Vehicles::OrderRingView(unk.orderOffset, 0).begin();
+                if (!order->hasFlag(Vehicles::OrderFlags::HasNumber))
+                {
+                    continue;
+                }
+
+                auto [loc, str] = sub_470B76(*order, i);
+                const auto pos = coordinate3dTo2d(loc.x, loc.y, loc.z, WindowManager::getCurrentRotation());
+                auto stringWidth = Gfx::getStringWidth(str);
+                for (auto zoom = 0; zoom < 4; ++zoom)
+                {
+                    // The first line of the label will always be at the centre
+                    // of the station/waypoint. This works out where the subsequent
+                    // lines of the label will end up.
+                    auto width = (stringWidth + 3) << zoom;
+                    auto numberHeight = (unk.lineNumber * 10) << zoom;
+                    auto firstLineHeight = 11 << zoom;
+                    auto midX = width / 2;
+                    auto midFirstLineY = firstLineHeight / 2;
+
+                    unk.position.left[zoom] = (pos.x - midX) >> zoom;
```

Comment:
position -> frame. A position is a 2d coordinate.

I'd prefer making a rect first, and then only at the end do the zoom transformation, but that would be weird with the string width/number height already being in the final scale...

What about first determining the x/y in the map scale, and then only doing the positioning in endsize coordinates?

Reply:
I don't see a better way that would get the same results.